apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,255 @@
+import type { database } from "@packages/db";
+import { libraryTraffic, libraryTrafficHistory } from "@packages/db/schema";
+import { conflictUpdateSetAllCols } from "@packages/db/utils";
+import { load } from "cheerio";
+import fetch from "cross-fetch";
+import { z } from "zod";
+
+const rawRespOKSchema = z.object({
+  message: z.literal("OK"),
+  data: z.object({
+    id: z.number(),
+    name: z.string(),
+    count: z.number(),
+    percentage: z.number(),
+    timestamp: z.string(),
+  }),
+});
+
+const rawRespSchema = z.union([
+  rawRespOKSchema,
+  z.object({ error: z.string() }),
+  z.object({ message: z.string() }),
+]);
+
+type RawRespOK = z.infer<typeof rawRespOKSchema>;
+
+export interface LocationMeta {
+  id: string;
+  libraryName: string;
+  locationLabel: string;
+  floorCode: string;
+}
+
+// Helper to sanitize and split raw JS array text content from HTML
+function parseArray(src: string): string[] {
+  return src
+    .split(/[,\n]/)
+    .map((s) => s.trim().replace(/^['"]|['"]$/g, ""))
+    .filter(Boolean);
+}
+
+// Build map of location metadata by ID from the UCI Libraries website
+async function collectLocationMeta(): Promise<Record<string, LocationMeta>> {
+  const html = await fetch("https://www.lib.uci.edu/where-do-you-want-study-today").then((r) =>
+    r.text(),
+  );
+  const $ = load(html);
+
+  const scriptText = $("script")
+    .toArray()
+    .map((el) => $(el).html() ?? "")
+    .find((txt) => /let\s+locationIds\s*=\s*\[/s.test(txt));
+
+  if (!scriptText) throw new Error("Could not find <script> containing locationIds");
+
+  const locMatch = scriptText.match(/let\s+locationIds\s*=\s*\[([\s\S]*?)]/);
+  const floorMatch = scriptText.match(/let\s+floors\s*=\s*\[([\s\S]*?)]/);
+  if (!locMatch || !floorMatch) throw new Error("Unable to capture locationIds or floors array");
+
+  // Extract parallel arrays:
+  // - locationIds: unique Occuspace IDs for each location (i.e. ['245', '677', '205'])
+  // - floors: corresponding internal DOM codes for each location's container (i.e. ['-GSC-2', '-SL-3', '1'])
+  const locationIds = parseArray(locMatch[1]);
+  const floorCodes = parseArray(floorMatch[1]);
+
+  if (locationIds.length !== floorCodes.length)
+    throw new Error("locationIds and floors arrays are different lengths");
+
+  const codeToLibrary = (code: string): string => {
```

Comment:
This isn't a function of anything in this body, so it should be a non-anonymous function at file scope.